File: packages/mcp/src/claude.ts
/**
 * Finding Claude on this machine, pairing it with this session, and opening it.
 *
 * The agent interface is only useful once something is on the other end of it,
 * and the last mile of that — "which client, told about which endpoint, and is
 * it running" — is the part a person is left to do by hand from a README. This
 * does it instead, from the one process that is in a position to: the
 * development server, which is on the same machine as the client and already
 * knows the endpoint's address.
 *
 * Everything it touches is behind `ClaudeEnvironment`. Not for neatness: the
 * whole of this file looks at the user's home directory, edits a configuration
 * file and starts a process, and none of that is something a test suite should
 * do to the machine it happens to be running on.
 */

/** Where the pairing is written, and what it is called there. */
export const CLAUDE_SERVER_NAME = 'panorama';

export interface ClaudeEnvironment {
  readonly platform: string;
  readonly home: string;
  /** The absolute path of a command on the PATH, or null. */
  find(command: string): Promise<string | null>;
  exists(path: string): Promise<boolean>;
  readFile(path: string): Promise<string | null>;
  writeFile(path: string, contents: string): Promise<void>;
  /** Runs a command to completion, and says how it went. */
  run(command: string, args: readonly string[]): Promise<{ code: number; output: string }>;
  /** Starts something and leaves it running, detached from this process. */
  start(command: string, args: readonly string[]): Promise<void>;
}

export interface ClaudeStatus {
  readonly platform: string;
  /** Claude Code, on the PATH. */
  readonly cli: { readonly found: boolean; readonly path?: string; readonly paired: boolean };
  /** The desktop application, and whether its configuration names this session. */
  readonly desktop: {
    readonly found: boolean;
    readonly configPath: string;
    readonly paired: boolean;
  };
  /** True when this machine is one we know how to open a terminal on. */
  readonly canOpenTerminal: boolean;
}

const DESKTOP_APP = '/Applications/Claude.app';

/**
 * Where the desktop application keeps its configuration.
 *
 * Only macOS and Windows have an answer; on anything else the desktop client
 * does not exist, and saying so is better than guessing at a path.
 */
export const desktopConfigPath = (environment: {
  readonly platform: string;
  readonly home: string;
}): string => {
  if (environment.platform === 'darwin') {
    return `${environment.home}/Library/Application Support/Claude/claude_desktop_config.json`;
  }
  if (environment.platform === 'win32') {
    return `${environment.home}/AppData/Roaming/Claude/claude_desktop_config.json`;
  }
  return '';
};

const readServers = async (
  environment: ClaudeEnvironment,
  path: string,
): Promise<Record<string, unknown>> => {
  if (path === '') return {};
  const contents = await environment.readFile(path);
  if (contents === null || contents.trim() === '') return {};
  try {
    const parsed = JSON.parse(contents) as Record<string, unknown>;
    const servers = parsed['mcpServers'];
    return typeof servers === 'object' && servers !== null
      ? (servers as Record<string, unknown>)
      : {};
  } catch {
    // A configuration file that cannot be read is not a configuration file this
    // has any business rewriting, and pairing will say so rather than replace it.
    return {};
  }
};

/** What is on this machine, and what already knows about this session. */
export const inspectClaude = async (environment: ClaudeEnvironment): Promise<ClaudeStatus> => {
  const cliPath = await environment.find('claude');
  const configPath = desktopConfigPath(environment);
  const [desktopFound, desktopServers, cliPaired] = await Promise.all([
    configPath === '' ? Promise.resolve(false) : environment.exists(DESKTOP_APP),
    readServers(environment, configPath),
    readServers(environment, `${environment.home}/.claude.json`),
  ]);
  return {
    platform: environment.platform,
    cli: {
      found: cliPath !== null,
      ...(cliPath === null ? {} : { path: cliPath }),
      paired: cliPaired[CLAUDE_SERVER_NAME] !== undefined,
    },
    desktop: {
      found: desktopFound || (configPath !== '' && (await environment.exists(configPath))),
      configPath,
      paired: desktopServers[CLAUDE_SERVER_NAME] !== undefined,
    },
    canOpenTerminal: environment.platform === 'darwin',
  };
};

export interface PairRequest {
  /** The endpoint an agent should talk to, e.g. `http://localhost:5173/agent/mcp`. */
  readonly url: string;
  /** Absolute path of the stdio pipe, for a client that speaks only that. */
  readonly bridgeScript: string;
}

export interface PairOutcome {
  readonly target: 'cli' | 'desktop';
  readonly done: boolean;
  readonly detail: string;
}

/**
 * Tells Claude Code about this session.
 *
 * Through its own command rather than by writing its file: the CLI owns that
 * schema, and a hand-written entry that is subtly wrong is worse than no entry —
 * it fails at the point where somebody is trying to use it, not here.
 */
const pairCli = async (
  environment: ClaudeEnvironment,
  request: PairRequest,
): Promise<PairOutcome> => {
  const found = await environment.find('claude');
  if (found === null) {
    return {
      target: 'cli',
      done: false,
      detail: 'Claude Code is not on the PATH of the process serving this page.',
    };
  }
  const result = await environment.run(found, [
    'mcp',
    'add',
    '--scope',
    'user',
    '--transport',
    'http',
    CLAUDE_SERVER_NAME,
    request.url,
  ]);
  return {
    target: 'cli',
    done: result.code === 0,
    detail:
      result.code === 0
        ? `Added "${CLAUDE_SERVER_NAME}" to Claude Code, for every project on this machine.`
        : `Claude Code refused: ${result.output.trim() || `exit ${result.code}`}`,
  };
};

/**
 * Tells the desktop application about this session.
 *
 * Written into its configuration, because it has no command to run — and merged
 * rather than replaced: it is the user's file, and whatever else is in it is
 * theirs. The entry is the stdio pipe, which is the transport that client
 * speaks.
 */
const pairDesktop = async (
  environment: ClaudeEnvironment,
  request: PairRequest,
): Promise<PairOutcome> => {
  const path = desktopConfigPath(environment);
  if (path === '') {
    return {
      target: 'desktop',
      done: false,
      detail: `There is no Claude desktop application on ${environment.platform}.`,
    };
  }
  const existing = await environment.readFile(path);
  if (existing !== null && existing.trim() !== '') {
    try {
      JSON.parse(existing);
    } catch {
      return {
        target: 'desktop',
        done: false,
        detail: `${path} is not valid JSON, so it has been left alone. Fix or move it and try again.`,
      };
    }
  }
  const parsed =
    existing === null || existing.trim() === ''
      ? {}
      : (JSON.parse(existing) as Record<string, unknown>);
  const servers = await readServers(environment, path);
  const contents = {
    ...parsed,
    mcpServers: {
      ...servers,
      [CLAUDE_SERVER_NAME]: {
        command: 'node',
        args: [request.bridgeScript],
        env: { PANORAMA_AGENT_URL: request.url },
      },
    },
  };
  await environment.writeFile(path, `${JSON.stringify(contents, null, 2)}\n`);
  return {
    target: 'desktop',
    done: true,
    detail: `Wrote "${CLAUDE_SERVER_NAME}" into ${path}. Restart Claude for it to be picked up.`,
  };
};

export const pairClaude = async (
  environment: ClaudeEnvironment,
  request: PairRequest,
  target: 'cli' | 'desktop' | 'both' = 'both',
): Promise<readonly PairOutcome[]> => {
  const outcomes: PairOutcome[] = [];
  if (target !== 'desktop') outcomes.push(await pairCli(environment, request));
  if (target !== 'cli') outcomes.push(await pairDesktop(environment, request));
  return outcomes;
};

export interface OpenOutcome {
  readonly opened: 'cli' | 'desktop' | null;
  readonly detail: string;
}

/**
 * How the desktop application is opened.
 *
 * Only ever asked where there is one to open, which is only ever the two
 * platforms it exists on — so this picks between them rather than guarding
 * against a third that `inspectClaude` has already ruled out.
 */
const desktopCommand = (platform: string): readonly [string, readonly string[]] =>
  platform === 'win32' ? ['cmd', ['/c', 'start', '', 'claude']] : ['open', ['-a', 'Claude']];

/**
 * Opens Claude — the application, if there is one.
 *
 * The application is what somebody who has installed it means by "open Claude":
 * it is a window that is already set up, and opening the command line instead
 * puts them in a terminal they did not ask for. So the desktop client is tried
 * first, and Claude Code is what there is when there is no application — opened
 * by asking the terminal to run it in the directory this project is being served
 * from, which is where it will find the project it is meant to be helping with.
 *
 * `prefer` overrides that in either direction, because a preference stated
 * outright should beat a default, and there are people for whom the terminal is
 * the point.
 */
export const openClaude = async (
  environment: ClaudeEnvironment,
  request: { readonly cwd: string; readonly prefer?: 'cli' | 'desktop' },
): Promise<OpenOutcome> => {
  const status = await inspectClaude(environment);
  const canOpenApp = status.desktop.found;
  const canOpenCli = status.cli.found && status.canOpenTerminal;
  const wanted = request.prefer ?? (canOpenApp ? 'desktop' : 'cli');
  // What was asked for, if it can be done, and otherwise whatever can.
  const target =
    wanted === 'desktop' && !canOpenApp
      ? 'cli'
      : wanted === 'cli' && !canOpenCli
        ? 'desktop'
        : wanted;

  if (target === 'desktop' && canOpenApp) {
    const [command, args] = desktopCommand(environment.platform);
    await environment.start(command, args);
    return { opened: 'desktop', detail: 'Opened the Claude desktop application.' };
  }
  if (target === 'cli' && canOpenCli) {
    // AppleScript, because a terminal is the one thing that cannot be started
    // headless: what makes Claude Code usable is having somewhere to type.
    const script = `tell application "Terminal" to do script "cd ${JSON.stringify(request.cwd).slice(1, -1)} && claude"\ntell application "Terminal" to activate`;
    await environment.start('osascript', ['-e', script]);
    return { opened: 'cli', detail: 'Opened Claude Code in a new Terminal window.' };
  }
  return {
    opened: null,
    detail:
      status.cli.found && !status.canOpenTerminal
        ? `Claude Code is installed, but opening a terminal is only implemented on macOS. Run "claude" in ${request.cwd}.`
        : 'Claude was not found on this machine.',
  };
};

File: apps/desktop/src-tauri/src/claude.rs
//! Finding Claude on this machine, telling it about this application, and opening
//! it — from inside the application, so that none of it needs a terminal.
//!
//! The development server does this too (`packages/mcp/src/claude.ts`), and the
//! reason it is here again rather than shared is that it is not the same job. That
//! one hands out an HTTP URL with a port in it; this one hands out **this
//! executable**, which is also the pipe (`--mcp-stdio`), and therefore never goes
//! stale: no port, no Node, nothing to reinstall when the application moves.
//!
//! One thing is harder here than there and worth knowing about. A development
//! server inherits a developer's shell, so `claude` is on its `PATH`. An
//! application launched from the Dock inherits almost nothing — `/usr/bin:/bin`
//! and little else — and Claude Code is usually installed somewhere that only a
//! login shell knows about. So the search asks the login shell as well, which is
//! what `find_claude` is mostly about.

use std::path::{Path, PathBuf};
use std::process::Command;

use serde::Serialize;

use crate::locate;

/// What the pairing is called in a client's configuration.
pub const SERVER_NAME: &str = "panorama";

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CliStatus {
    pub found: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
    pub paired: bool,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopStatus {
    pub found: bool,
    pub config_path: String,
    pub paired: bool,
}

/// The shape the settings panel already reads, so the panel needs no idea that it
/// is talking to a shell rather than to a development server.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Status {
    pub platform: String,
    pub cli: CliStatus,
    pub desktop: DesktopStatus,
    pub can_open_terminal: bool,
    pub mcp_url: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PairOutcome {
    pub target: String,
    pub done: bool,
    pub detail: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenOutcome {
    pub opened: Option<String>,
    pub detail: String,
}

/// Where Claude Code might be, beyond the usual places: its own installer puts it
/// here.
fn likely_paths() -> Vec<PathBuf> {
    let mut places = vec![locate::home().join(".claude/local/claude")];
    places.extend(locate::usual_places("claude"));
    places
}

/// `claude`, wherever it is. See `locate.rs` for why this is not one line.
pub fn find_claude() -> Option<PathBuf> {
    locate::find("claude", &likely_paths())
}

/// Where the desktop application keeps its configuration. Empty where there is no
/// such application, which is every platform but two.
pub fn desktop_config_path() -> PathBuf {
    if cfg!(target_os = "macos") {
        locate::home().join("Library/Application Support/Claude/claude_desktop_config.json")
    } else if cfg!(windows) {
        locate::home().join("AppData/Roaming/Claude/claude_desktop_config.json")
    } else {
        PathBuf::new()
    }
}

/// Whether a configuration file already names this application.
///
/// Reads `mcpServers` out of whatever JSON is there; a file that cannot be parsed
/// is not one this has any business having an opinion about.
pub fn names_us(text: &str) -> bool {
    serde_json::from_str::<serde_json::Value>(text)
        .ok()
        .and_then(|value| value.get("mcpServers")?.get(SERVER_NAME).cloned())
        .is_some()
}

fn read(path: &Path) -> Option<String> {
    std::fs::read_to_string(path).ok()
}

/// This application, as a client would have to start it.
fn pipe_command() -> (String, Vec<String>) {
    let exe = std::env::current_exe()
        .map(|path| path.to_string_lossy().to_string())
        .unwrap_or_else(|_| "panorama-desktop".to_string());
    (exe, vec!["--mcp-stdio".to_string()])
}

pub fn status(mcp_url: String) -> Status {
    let cli = find_claude();
    let config = desktop_config_path();
    let desktop_config = if config.as_os_str().is_empty() {
        None
    } else {
        read(&config)
    };
    Status {
        platform: std::env::consts::OS.to_string(),
        cli: CliStatus {
            found: cli.is_some(),
            path: cli.as_ref().map(|path| path.to_string_lossy().to_string()),
            paired: read(&locate::home().join(".claude.json"))
                .map(|text| names_us(&text))
                .unwrap_or(false),
        },
        desktop: DesktopStatus {
            // The application on macOS, or a configuration file it left behind on
            // either platform — which is the only evidence there is on Windows.
            found: Path::new("/Applications/Claude.app").exists() || desktop_config.is_some(),
            config_path: config.to_string_lossy().to_string(),
            paired: desktop_config.as_deref().map(names_us).unwrap_or(false),
        },
        can_open_terminal: cfg!(target_os = "macos"),
        mcp_url,
    }
}

/// Tells Claude Code about this application, through its own command rather than
/// by writing its file: the CLI owns that schema, and an entry that is subtly
/// wrong fails where somebody is trying to use it rather than here.
fn pair_cli() -> PairOutcome {
    let Some(claude) = find_claude() else {
        return PairOutcome {
            target: "cli".to_string(),
            done: false,
            detail: "Claude Code was not found on this machine.".to_string(),
        };
    };
    let (command, args) = pipe_command();
    // Through `locate::command`, because on Windows the CLI is often a `.cmd`
    // shim and those cannot be executed directly.
    let mut invocation = locate::command(&claude);
    invocation.args(["mcp", "add", "--scope", "user", SERVER_NAME, "--"]);
    invocation.arg(&command);
    invocation.args(&args);
    match invocation.output() {
        Ok(result) if result.status.success() => PairOutcome {
            target: "cli".to_string(),
            done: true,
            detail: format!(
                "Added \"{SERVER_NAME}\" to Claude Code, for every project on this machine."
            ),
        },
        Ok(result) => {
            let said = String::from_utf8_lossy(&result.stderr);
            let said = if said.trim().is_empty() {
                String::from_utf8_lossy(&result.stdout).trim().to_string()
            } else {
                said.trim().to_string()
            };
            PairOutcome {
                target: "cli".to_string(),
                done: false,
                detail: format!("Claude Code refused: {said}"),
            }
        }
        Err(problem) => PairOutcome {
            target: "cli".to_string(),
            done: false,
            detail: format!("Could not run {}: {problem}", claude.to_string_lossy()),
        },
    }
}

/// The configuration a desktop client needs, merged into whatever is already in
/// the file. It is the user's file; everything else in it is theirs.
pub fn merge_into(
    existing: Option<&str>,
    command: &str,
    args: &[String],
) -> Result<String, String> {
    let mut root = match existing {
        None => serde_json::json!({}),
        Some(text) if text.trim().is_empty() => serde_json::json!({}),
        Some(text) => serde_json::from_str::<serde_json::Value>(text)
            .map_err(|_| "is not valid JSON, so it has been left alone".to_string())?,
    };
    if !root.is_object() {
        return Err("does not hold a JSON object, so it has been left alone".to_string());
    }
    let entry = serde_json::json!({ "command": command, "args": args });
    let object = root.as_object_mut().expect("checked above");
    let servers = object
        .entry("mcpServers")
        .or_insert_with(|| serde_json::json!({}));
    if !servers.is_object() {
        *servers = serde_json::json!({});
    }
    servers
        .as_object_mut()
        .expect("just made one")
        .insert(SERVER_NAME.to_string(), entry);
    Ok(format!(
        "{}\n",
        serde_json::to_string_pretty(&root).map_err(|problem| problem.to_string())?
    ))
}

fn pair_desktop() -> PairOutcome {
    let path = desktop_config_path();
    if path.as_os_str().is_empty() {
        return PairOutcome {
            target: "desktop".to_string(),
            done: false,
            detail: format!(
                "There is no Claude desktop application on {}.",
                std::env::consts::OS
            ),
        };
    }
    let (command, args) = pipe_command();
    match merge_into(read(&path).as_deref(), &command, &args) {
        Err(problem) => PairOutcome {
            target: "desktop".to_string(),
            done: false,
            detail: format!(
                "{} {problem}. Fix or move it and try again.",
                path.display()
            ),
        },
        Ok(contents) => {
            if let Some(parent) = path.parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            match std::fs::write(&path, contents) {
                Ok(()) => PairOutcome {
                    target: "desktop".to_string(),
                    done: true,
                    detail: format!(
                        "Wrote \"{SERVER_NAME}\" into {}. Restart Claude for it to be picked up.",
                        path.display()
                    ),
                },
                Err(problem) => PairOutcome {
                    target: "desktop".to_string(),
                    done: false,
                    detail: format!("Could not write {}: {problem}", path.display()),
                },
            }
        }
    }
}

pub fn pair() -> Vec<PairOutcome> {
    vec![pair_cli(), pair_desktop()]
}

/// Opens Claude — the application where there is one, because that is what
/// somebody who installed it means by "open Claude"; the command line otherwise,
/// in a terminal, because what makes Claude Code usable is having somewhere to
/// type.
pub fn open(prefer: Option<&str>) -> OpenOutcome {
    let status = status(String::new());
    let can_app = status.desktop.found && (cfg!(target_os = "macos") || cfg!(windows));
    let can_cli = status.cli.found && status.can_open_terminal;
    let wanted = prefer.unwrap_or(if can_app { "desktop" } else { "cli" });
    let target = match wanted {
        "desktop" if !can_app => "cli",
        "cli" if !can_cli => "desktop",
        other => other,
    };

    if target == "desktop" && can_app {
        let started = if cfg!(windows) {
            Command::new("cmd")
                .args(["/c", "start", "", "claude"])
                .spawn()
        } else {
            Command::new("open").args(["-a", "Claude"]).spawn()
        };
        return match started {
            Ok(_) => OpenOutcome {
                opened: Some("desktop".to_string()),
                detail: "Opened the Claude desktop application.".to_string(),
            },
            Err(problem) => OpenOutcome {
                opened: None,
                detail: format!("Could not open the Claude application: {problem}"),
            },
        };
    }
    if target == "cli" && can_cli {
        // AppleScript, because a terminal is the one thing that cannot be started
        // headless. The directory is the user's own: this application is not in a
        // project, and inventing one would put Claude somewhere arbitrary.
        let where_to = locate::home().to_string_lossy().to_string();
        let script = format!(
            "tell application \"Terminal\" to do script \"cd {where_to} && claude\"\ntell application \"Terminal\" to activate"
        );
        return match Command::new("osascript").args(["-e", &script]).spawn() {
            Ok(_) => OpenOutcome {
                opened: Some("cli".to_string()),
                detail: format!("Opened Claude Code in a new Terminal window, in {where_to}."),
            },
            Err(problem) => OpenOutcome {
                opened: None,
                detail: format!("Could not open a terminal: {problem}"),
            },
        };
    }
    OpenOutcome {
        opened: None,
        detail: if status.cli.found && !status.can_open_terminal {
            "Claude Code is installed, but opening a terminal is only implemented on macOS."
                .to_string()
        } else {
            "Claude was not found on this machine.".to_string()
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn recognises_a_configuration_that_already_names_us() {
        assert!(names_us(
            "{\"mcpServers\":{\"panorama\":{\"command\":\"x\"}}}"
        ));
        assert!(!names_us("{\"mcpServers\":{\"other\":{}}}"));
        assert!(!names_us("{}"));
        // A file that is not JSON is not one to have an opinion about.
        assert!(!names_us("nonsense"));
    }

    #[test]
    fn merges_into_a_file_that_is_somebody_elses() {
        let existing = "{\n  \"theme\": \"dark\",\n  \"mcpServers\": {\n    \"other\": { \"command\": \"o\" }\n  }\n}";
        let merged = merge_into(
            Some(existing),
            "/Applications/P.app/x",
            &["--mcp-stdio".to_string()],
        )
        .expect("valid JSON");
        let parsed: serde_json::Value = serde_json::from_str(&merged).expect("wrote JSON");
        // Everything that was there is still there.
        assert_eq!(parsed["theme"], "dark");
        assert_eq!(parsed["mcpServers"]["other"]["command"], "o");
        // And this application is now in it, as the pipe.
        assert_eq!(
            parsed["mcpServers"]["panorama"]["command"],
            "/Applications/P.app/x"
        );
        assert_eq!(parsed["mcpServers"]["panorama"]["args"][0], "--mcp-stdio");
    }

    #[test]
    fn writes_a_whole_file_where_there_was_none() {
        let merged = merge_into(None, "x", &[]).expect("no file is not an error");
        assert!(merged.contains("\"panorama\""));
        assert!(merged.ends_with('\n'));
        assert_eq!(merge_into(Some("   "), "x", &[]).unwrap(), merged);
    }

    #[test]
    fn refuses_to_touch_a_file_it_cannot_read() {
        // Better than replacing it: the file is the user's, and a configuration
        // that was hand-edited into invalidity is something they will want back.
        let refused = merge_into(Some("{ not json"), "x", &[]).expect_err("should refuse");
        assert!(refused.contains("not valid JSON"));
        assert!(merge_into(Some("[1,2]"), "x", &[]).is_err());
    }

    #[test]
    fn replaces_an_mcpservers_that_is_not_an_object() {
        // Seen in the wild as `"mcpServers": null` from a half-written edit.
        let merged = merge_into(Some("{\"mcpServers\": null}"), "x", &[]).expect("recoverable");
        assert!(names_us(&merged));
    }

    /// Not part of the suite: it needs Claude Code installed on the machine
    /// running it, which CI does not have.
    ///
    /// Worth running by hand after touching the search — `cargo test -- --ignored`
    /// — because it is the one thing here that fails *silently* in a real
    /// application: launched from the Dock, this process inherits almost no `PATH`,
    /// and a search that only reads `PATH` finds nothing while working perfectly
    /// in every test and every terminal.
    #[test]
    #[ignore]
    fn finds_claude_with_the_path_an_application_actually_has() {
        let restore = std::env::var("PATH").ok();
        std::env::set_var("PATH", "/usr/bin:/bin");
        let found = find_claude();
        match restore {
            Some(path) => std::env::set_var("PATH", path),
            None => std::env::remove_var("PATH"),
        }
        assert!(
            found.is_some(),
            "neither the likely places nor the login shell found claude"
        );
    }

    #[test]
    fn pairs_with_this_executable_rather_than_a_port() {
        let (command, args) = pipe_command();
        assert!(!command.is_empty());
        assert_eq!(args, vec!["--mcp-stdio".to_string()]);
    }
}
